crates/telio-wg/src/wg.rs

```diff
@@ -124,9 +124,75 @@ pub struct Io {
 pub const KEEPALIVE_PACKET_SIZE: u64 = 32;
 /// Default wireguard keepalive duration
 pub const WG_KEEPALIVE: Duration = Duration::from_secs(10);
+/// Used to ensure uapi errors happen for at least this long before considering critical error event.
+/// The failures are usually transient and observed during power transitions when using WireguardNT.
+const REQUIRED_CRITICAL_ERROR_GAP: Duration = Duration::from_secs(15);
+/// Used to detect long gaps between UAPI commands that suggest the device might have been suspended.
+/// Usually libtelio should poll every second.
+const MAX_UAPI_CMD_GAP: Duration = Duration::from_secs(30);
 
 pub use telio_utils::{BytesAndTimestamps, Instant};
 
+#[derive(Debug)]
+struct UAPIErrorTracker {
+    // Minimum required time between first error and later ones to consider returning critical verdict.
+    required_critical_error_gap: Duration,
+
+    // Maximum time allowed between uapi commands.
+    max_uapi_cmd_gap: Duration,
+
+    // Timestamp of the first error from which the required_critical_error_gap is measured.
+    first_error_at: Option<Instant>,
+
+    // Timestamp of the most recent UAPI command.
+    last_uapi_cmd_at: Option<Instant>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum UAPIErrorVerdict {
+    Ignore,
+    Critical,
+}
+
+impl UAPIErrorTracker {
+    fn new(required_critical_error_gap: Duration, max_uapi_cmd_gap: Duration) -> Self {
+        UAPIErrorTracker {
+            required_critical_error_gap,
+            max_uapi_cmd_gap,
+            first_error_at: None,
+            last_uapi_cmd_at: None,
+        }
+    }
+
+    fn consume_response(&mut self, errno: i32, inst: Instant) -> UAPIErrorVerdict {
```

Comment:
Please consider a more descriptive name for this method, perhaps `process_uapi_response()` or `evaluate_uapi_response()`. The current name "consume" doesn't tell me what to expect from it. Would be nice if it conveyed that it tracks error state over time and returns a verdict.